### PR TITLE
Fix nested objects undo

### PR DESCRIPTION
**Pull request by lewisjb**
### Context
This change solves the issue of undoing a `remove_row` while using a nested object data source not bringing data back (jsfiddle and animation can be found at #3958). It is needed because undoing is a very helpful feature and should work properly for all supported types of data sources.

It is worth noting that the changes to the undo/redo plugin are required to bring back hidden data. The other changes bring back visible data.

### How has this been tested?
I have tested that it fixes the issue by using it repeatedly. I have also tested it on some of the demos to ensure that my fix didn't break other things.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Related issue(s):
1. #3958 
2. #3959

### Checklist:
- [x] My code follows the code style of this project,
- [x] Added relevant tests
- [ ] My change requires a change to the documentation.

### CODE REVIEW

**Comment by AMBudnik** [comment]
Hi @lewisjb 

Thank you for your PR and a great description. I see that you did not finish all the tasks that you wanted to do so I just wanted to let you know that (by following our lawyers advice) we need to get a signed permission for code us from all our future contributors. We are waiting for the document to be finished so I wanted to ask if you will be willing to sign it?

**Comment by lewisjb** [comment]
Hi @AMBudnik

Thanks for your response. I am not sure what you mean by "did not finish all the tasks"; if you are referring to the unchecked 'documentation' box, that is from the pull request template.
The comment above the checklist states "put an `x` in all the boxes that apply" (see: [template](https://github.com/handsontable/handsontable/blame/master/.github/PULL_REQUEST_TEMPLATE.md#L19)), since my pull request is a bug-fix, I don't believe it requires a documentation change.

Regarding the document, I would be happy to get a contributor license agreement signed by my employer, who currently owns the intellectual property.

**Comment by AMBudnik** [comment]
Thank you for your prompt reply @lewisjb 

I will send the document/or link as soon as possible.

**Comment by lewisjb** [comment]
@AMBudnik Is there an ETA on the CLA?

**Comment by AMBudnik** [comment]
@lewisjb 
Sorry for keeping you waiting for so long. It should be ready for the newest release which is planned for tomorrow. The change relates to the whole Contributing file so we are waiting for the release.

**Comment by AMBudnik** [comment]
Hi @lewisjb 

the changes are published. The new Contributing file is available [here](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md).

As most of the PR in our repository are 3 months+ old we have decided to close them all and reopen immediately after the author confirms that all the steps of the [Contributing file](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) have been accomplished. 

ps. if you have added more than one pull request please share all their numbers in one post. Thanks in advance!

**Comment by lewisjb** [comment]
@AMBudnik Done.

**Comment by AMBudnik** [comment]
Thank you @lewisjb 

I am reopening the ticket.

**Comment by TRManderson** [comment]
Is there a timeline for how long we should expect before this PR can be reviewed? (and thus merged)

**Comment by AMBudnik** [comment]
@TRManderson 

The bigger the change is the longer it gets to test it. 

The quality of tests is also an important factor. 

Usually, each Friday in our company is called a _Review day_, which means that our devs review code of their colleagues and if there's a time also the code of our contributors.

Please note that the PR with a label `Developed (ready for review)` is the last label that is added to the pull request before a developer can test it.

**Comment by AMBudnik** [comment]
ps. it is forbidden to commit `/dist` files. It's in the point 4 of the [Contributing guidelines](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md).

Also, the undo/redo plugin is qualified to be refactored so the code will change.

**Comment by lewisjb** [comment]
Apologies for the modification of `/dist`, my co-worker messed up the branch. If fixing that won't get the branch merged anyway due to undo/redo being refactored, I won't bother fixing it; if that's not the case, I will fix the branch ASAP.

Do you have an ETA on the refactor? And will the bug that this PR is trying to fix be fixed in it?

Regarding the response about how HoT reviews PRs, this PR was a fairly small change, and I'd like to this the tests were up to your standards as I based them off your existing ones.

**Comment by AMBudnik** [comment]
That's OK @lewisjb it's hard to control all the co-workers. I know that you're an experienced Handsontable user. I have seen you in back in 2016 posting an issue.

We do not have the refactor scheduled yet, however, it would be your waste of time to prepare a solution if it won't be merged. If a plugin is qualified to me refactored we usually do not fix smaller issues one by one, rather take some time to refactor the plugin and then fix the issues.

**Comment by lewisjb** [comment]
Has the refactor been scheduled yet?

**Comment by AMBudnik** [comment]
The refactor has been skipped in our [Roadmap for this year](https://trello.com/b/PztR4hpj/handsontable-roadmap).

**Comment by TRManderson** [comment]
Given the refactor has been skipped in your 2018 roadmap, can you please reopen and merge this pull request?

**Comment by AMBudnik** [comment]
Sure. I can open the pull request and add a 'ready for review' label if it still follows our contribution rules.

**Comment by lewisjb** [comment]
@AMBudnik I have cleaned up the branch, and the tests are passing; so I believe it is ready for review.

**Comment by AMBudnik** [comment]
Thank you @lewisjb 
I am returning the label.

File: src/plugins/undoRedo/test/UndoRedo.e2e.js [hunk 1]
```diff
@@ -2090,6 +2090,291 @@ describe('UndoRedo', () => {
...
+        return [
+          {name: {first: 'Timothy', last: 'Dalton'}, year: 1986},
+          {name: {first: 'Sean', last: 'Connery'}, year: 1962},
+          {name: {first: 'Roger', last: 'Moore'}, year: 1973}
+        ];
+      }
+
+      describe('undo', () => {
+        it('should undo single change', () => {
+          handsontable({
```
Comment (swistach) [inline src/plugins/undoRedo/test/UndoRedo.e2e.js]: You can use `const hot = handsontable(...)` instead.

File: src/plugins/undoRedo/test/UndoRedo.e2e.js [hunk 2]
```diff
@@ -2090,6 +2090,291 @@ describe('UndoRedo', () => {
...
+          expect(getDataAtRowProp(1, 'name.first')).toEqual('Sean');
+          expect(getDataAtRowProp(1, 'name.last')).toEqual('Connery');
+          expect(getDataAtRowProp(1, 'year')).toEqual(1962);
+          expect(getDataAtRowProp(2, 'name.first')).toEqual('Roger');
+          expect(getDataAtRowProp(2, 'name.last')).toEqual('Moore');
+          expect(getDataAtRowProp(2, 'year')).toEqual(1973);
+        });
+
+        it('should undo removal of hidden columns', () => {
+          var HOT = handsontable({
```
Comment (swistach) [inline src/plugins/undoRedo/test/UndoRedo.e2e.js]: Could you use `const` instead of `var`?

File: src/plugins/undoRedo/test/UndoRedo.e2e.js [hunk 3]
```diff
@@ -2090,6 +2090,291 @@ describe('UndoRedo', () => {
...
+
+          HOT.undo();
+          expect(getDataAtCell(0, 0)).toBe('Timothy');
+
+          HOT.redo();
+          expect(getDataAtCell(0, 0)).toBe('Pearce');
+        });
+
+        it('should redo removal of single row', () => {
+          var HOT = handsontable({
```
Comment (swistach) [inline src/plugins/undoRedo/test/UndoRedo.e2e.js]: Could you use `const` instead of `var`?

File: src/plugins/undoRedo/test/UndoRedo.e2e.js [hunk 4]
```diff
@@ -2090,6 +2090,291 @@ describe('UndoRedo', () => {
...
+          expect(getDataAtRowProp(0, 'name.first')).toEqual('Timothy');
+          expect(getDataAtRowProp(0, 'name.last')).toEqual('Dalton');
+          expect(getDataAtRowProp(0, 'year')).toEqual(1986);
+          expect(getDataAtRowProp(1, 'name.first')).toBeNull();
+          expect(getDataAtRowProp(1, 'name.last')).toBeNull();
+          expect(getDataAtRowProp(1, 'year')).toBeNull();
+        });
+
+        it('should redo removal of multiple rows', () => {
+          var HOT = handsontable({
```
Comment (swistach) [inline src/plugins/undoRedo/test/UndoRedo.e2e.js]: Could you use `const` instead of `var`?

File: src/plugins/undoRedo/test/UndoRedo.e2e.js [hunk 5]
```diff
@@ -2090,6 +2090,291 @@ describe('UndoRedo', () => {
...
+          expect(getDataAtRowProp(1, 'name.first')).toBeNull();
+          expect(getDataAtRowProp(1, 'name.last')).toBeNull();
+          expect(getDataAtRowProp(1, 'year')).toBeNull();
+          expect(getDataAtRowProp(2, 'name.first')).toBeNull();
+          expect(getDataAtRowProp(2, 'name.last')).toBeNull();
+          expect(getDataAtRowProp(2, 'year')).toBeNull();
+        });
+
+        it('should redo removal of hidden columns', () => {
+          var HOT = handsontable({
```
Comment (swistach) [inline src/plugins/undoRedo/test/UndoRedo.e2e.js]: Could you use `const` instead of `var`?

File: src/plugins/undoRedo/undoRedo.js [hunk 6]
```diff
@@ -49,7 +49,7 @@ function UndoRedo(instance) {
       return;
     }
 
-    var originalData = plugin.instance.getSourceDataArray();
+    var originalData = plugin.instance.getSourceData();
```
Comment (swistach) [inline src/plugins/undoRedo/undoRedo.js]: Storing a reference to source data isn't the best choice.
Maybe you can find a different way (without storing the reference) to save removed data?

File: src/plugins/undoRedo/undoRedo.js [hunk 7]
```diff
@@ -315,9 +315,15 @@ UndoRedo.RemoveRowAction = function(index, data) {
 inherit(UndoRedo.RemoveRowAction, UndoRedo.Action);
 
 UndoRedo.RemoveRowAction.prototype.undo = function(instance, undoneCallback) {
-  instance.alter('insert_row', this.index, this.data.length, 'UndoRedo.undo');
   instance.addHookOnce('afterRender', undoneCallback);
-  instance.populateFromArray(this.index, 0, this.data, void 0, void 0, 'UndoRedo.undo');
+  instance.alter('insert_row', this.index, this.data.length, 'undo');
+
+  var source = instance.getSourceData();
```
Comment (swistach) [inline src/plugins/undoRedo/undoRedo.js]: It's a breaking change. Do you think it’s a better option, to change data silently (without calling any hooks)?

File: src/dataSource.js [hunk 8]
```diff
@@ -45,7 +45,7 @@ class DataSource {
     if (toArray) {
       result = this.getByRange(
         {row: 0, col: 0},
-        {row: Math.max(this.countRows() - 1, 0), col: Math.max(this.countColumns() - 1, 0)},
+        {row: Math.max(this.countRows() - 1, 0), col: Math.max(this.hot.countCols() - 1, 0)},
```
[the remark this answers is not in the document]
Comment (lewisjb) [reply src/dataSource.js]: Unless I'm mistaken, in #3958 you said `this.countColumns` was fine and I should use `hot.countCols`

File: src/plugins/undoRedo/undoRedo.js (hunk 6 above)
Comment (lewisjb) [reply src/plugins/undoRedo/undoRedo.js]: It already used `getSourceDataArray`, what's the difference?

File: src/dataSource.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (swistach) [reply src/dataSource.js]: To be honest, I was wrong.
Currently, I see it as below:

DataMap and DataSource should have their own `count...` methods.
`Core.count...` methods should use `datamap` instance.
`Core.countSource...` methods should use `datasource` instance.

But first I would like to discuss this change with rest of Core team.

File: src/plugins/undoRedo/undoRedo.js (hunk 6 above)
Comment (swistach) [reply src/plugins/undoRedo/undoRedo.js]: `.getSourceData()` returns reference to source data:
https://github.com/handsontable/handsontable/blob/master/src/core.js#L2085
then
https://github.com/handsontable/handsontable/blob/master/src/dataSource.js#L42
`toArray` is false, so `getData()` return `dataSource.data`

but `getSourceDataArray()` returns a new array with values from source data, there is no reference, or maybe better, shouldn't keep any reference to source data.
https://github.com/handsontable/handsontable/blob/master/src/core.js#L2110
then 
https://github.com/handsontable/handsontable/blob/master/src/dataSource.js#L42
but this time `toArray` is true, so our result is `datasource.getByRange(start, end, toArray)`

I hope my explanation is clear enough. Let me know it isn't.

Comment (lewisjb) [reply src/plugins/undoRedo/undoRedo.js]: On line 56 it does a `deepClone`, which does a JSON deep-clone; so it does a complete deep copy of the data; so no reference to the initial data is exposed.

Comment (swistach) [reply src/plugins/undoRedo/undoRedo.js]: You're right. I didn't notice hidden part of code on a PR preview.

File: src/plugins/undoRedo/undoRedo.js (hunk 7 above)
Comment (AMBudnik) [reply src/plugins/undoRedo/undoRedo.js]: @lewisjb please help us to pull this PR. It has been a very long time since our last conversation.

Comment (lewisjb) [reply src/plugins/undoRedo/undoRedo.js]: I was waiting for an update on https://github.com/handsontable/handsontable/pull/3998#discussion_r190837441

Comment (AMBudnik) [reply src/plugins/undoRedo/undoRedo.js]: Sorry, I haven't seen that comment. I've asked @swistach to send his reply as soon as possible

File: src/dataSource.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (swistach) [reply src/dataSource.js]: I can confirm, `this.countColumns()` must stay.

Maybe problem is in countColumn which doesn't count object's keys on each level.

File: src/plugins/undoRedo/undoRedo.js (hunk 7 above)
Comment (lewisjb) [reply src/plugins/undoRedo/undoRedo.js]: I think it's probably better to avoid a breaking change for this

**Comment by lewisjb** [comment]
I've updated the PR to resolve the feedback. I'm not a huge fan of: https://github.com/handsontable/handsontable/pull/3998/files#diff-1077d5ee9817a08cce35d56e6e0cbe07R343 but something was needed to be able to call `populateFromArray` to avoid a breaking change (and breaking Formula tests)

**Review by wszymanski — COMMENTED** [review]
Hello, @lewisjb. I'm sorry that you has waited so long. It's nice that you have tried not to do a breaking change. It seems that #3958 and #3959 were fixed.  But... tests created by you shows that we we still have problem for restoring data for columns which haven't been presented in dataset (called `hidden` by you).

However, the solution provided within PR won't be useful anymore as https://github.com/handsontable/handsontable/pull/6678 changed what `getSourceData` returns. There is no possibility to use that code.

**Comment by lewisjb** [comment]
It's a shame to see that this code is now outdated but the bug from 4+ years ago is still present when there was a fix sitting here.
Due to bugs like this and lack of support for them, we stopped using this software years ago and wrote our own as it would be less effort.
I probably won't fix this PR considering we no longer use this, and whenever I've updated this PR in the past it still hasn't been merged.

**Comment by AMBudnik** [comment]
Thank you for your effort here and before, @lewisjb 
We will no longer message you about this pull request. At the same time, I'm relieved to know that you have a working solution. 

@wszymanski I have created an issue for us to fix based on the test scenario https://github.com/handsontable/handsontable/issues/8039

**Pull request closed without merging**